fem/libceed/ceed.cpp

```diff
@@ -60,8 +62,201 @@ void InitCeedCoeff(Coefficient* Q, CeedData* ptr)
+                                    &ceedData.restr_i);
+
+   CeedVectorCreate(ceed, mesh->GetNodes()->Size(), &ceedData.node_coords);
+   CeedVectorSetArray(ceedData.node_coords, CEED_MEM_HOST, CEED_USE_POINTER,
+                      mesh->GetNodes()->GetData());
+
+   CeedVectorCreate(ceed, nelem * nqpts * qdatasize, &ceedData.rho);
+
+   // Context data to be passed to the 'f_build_diff' Q-function.
+   ceedData.build_ctx_data.dim = mesh->Dimension();
+   ceedData.build_ctx_data.space_dim = mesh->SpaceDimension();
+   ceedData.build_ctx_data.vdim = fes.GetVDim();
+
+   std::string qf_file = GetCeedPath() + op.header;
+   std::string qf;
+
+   // Create the Q-function that builds the operator (i.e. computes its
+   // quadrature data) and set its context data.
+   switch (ceedData.coeff_type)
+   {
+      case CeedCoeff::Const:
+         qf = qf_file + op.const_func;
+         CeedQFunctionCreateInterior(ceed, 1, op.const_qf,
+                                     qf.c_str(),
+                                     &ceedData.build_qfunc);
+         ceedData.build_ctx_data.coeff = ((CeedConstCoeff*)ceedData.coeff)->val;
+         break;
+      case CeedCoeff::Grid:
+         qf = qf_file + op.grid_func;
+         CeedQFunctionCreateInterior(ceed, 1, op.grid_qf,
+                                     qf.c_str(),
+                                     &ceedData.build_qfunc);
+         CeedQFunctionAddInput(ceedData.build_qfunc, "coeff", 1, CEED_EVAL_INTERP);
+         break;
+      default:
+         MFEM_ABORT("This coeff_type is not handled");
+   }
+   CeedQFunctionAddInput(ceedData.build_qfunc, "dx", dim * dim, CEED_EVAL_GRAD);
+   CeedQFunctionAddInput(ceedData.build_qfunc, "weights", 1, CEED_EVAL_WEIGHT);
+   CeedQFunctionAddOutput(ceedData.build_qfunc, "qdata", qdatasize,
+                          CEED_EVAL_NONE);
+
+   CeedQFunctionContextCreate(ceed, &ceedData.build_ctx);
+   CeedQFunctionContextSetData(ceedData.build_ctx, CEED_MEM_HOST, CEED_USE_POINTER,
+                               sizeof(ceedData.build_ctx_data),
+                               &ceedData.build_ctx_data);
+   CeedQFunctionSetContext(ceedData.build_qfunc, ceedData.build_ctx);
+
+   // Create the operator that builds the quadrature data for the operator.
+   CeedOperatorCreate(ceed, ceedData.build_qfunc, NULL, NULL,
+                      &ceedData.build_oper);
+   if (ceedData.coeff_type==CeedCoeff::Grid)
+   {
+      CeedGridCoeff* ceedCoeff = (CeedGridCoeff*)ceedData.coeff;
+      InitCeedBasisAndRestriction(*ceedCoeff->coeff->FESpace(), irm, ceed,
+                                  &ceedCoeff->basis,
+                                  &ceedCoeff->restr);
+      CeedVectorCreate(ceed, ceedCoeff->coeff->FESpace()->GetNDofs(),
+                       &ceedCoeff->coeffVector);
+      CeedVectorSetArray(ceedCoeff->coeffVector, CEED_MEM_HOST, CEED_USE_POINTER,
+                         ceedCoeff->coeff->GetData());
```

Comment:
Possible optimization: avoid host-to-device copy by using the device pointer from `*ceedCoeff->coeff`.

Reply:
This optimization is already done in `test-ceed` is it okay if I leave this one like this here? I also moved the `CeedVector` definition into `InitCeedCoeff`.

Reply:
```c++
   else if (GridFunctionCoefficient* coeff =
               dynamic_cast<GridFunctionCoefficient*>(Q))
   {
      CeedGridCoeff *ceedCoeff = new CeedGridCoeff;
      ceedCoeff->coeff = coeff->GetGridFunction();
      CeedVectorCreate(internal::ceed, ceedCoeff->coeff->FESpace()->GetNDofs(),
                       &ceedCoeff->coeffVector);
      CeedScalar *d_ptr;
      CeedMemType mem;
      CeedGetPreferredMemType(internal::ceed, &mem);
      if ( Device::Allows(Backend::CUDA) && mem==CEED_MEM_DEVICE )
      {
         d_ptr = const_cast<CeedScalar*>(ceedCoeff->coeff->Read());
      }
      else
      {
         d_ptr = const_cast<CeedScalar*>(ceedCoeff->coeff->HostRead());
         mem = CEED_MEM_HOST;
      }
      CeedVectorSetArray(ceedCoeff->coeffVector, mem, CEED_USE_POINTER,
                         d_ptr);
      ptr->coeff_type = CeedCoeff::Grid;
      ptr->coeff = static_cast<void*>(ceedCoeff);
   }
```

Reply:
Okay. You can also update `Backend::CUDA` to `Backend::DEVICE_MASK` there.

In the future, this way of reviewing and merging PRs should definitely be avoided. The burden of tracking the changes from this PR into `test-ceed` (#1719) is asking too much from the reviewers.

If there are big conflicts, then some of the PRs will wait for the conflicting ones to be merged in `master` first, so that the author can resolve the conflicts in his PR before merging into `next`.